### DISCUSSION TITLE
Integrate with an app containing rate limiter

### DISCUSSION THREAD

**Discussion by GingerMoon** [dotnet/yarp#1854] [Q&A] [answered]
Hi guys, 

We are migrating a legacy .net framework app to .net6. 
We want to use yarp to switch the traffic smoothly.
However, this app contains its own rate limiter, which works before the Endpoints middleware. 
According to my understanding, yarp works as a sub-pipeline inside the Endpoints middleware correct?
I am wondering if this lib supports bringing forward the reverse proxy pipeline (at least ahead of the endpoints middleware)?
This topic is a little similiar as https://github.com/microsoft/reverse-proxy/discussions/1464

**Comment by samsp-msft** [comment]
I am assuming that what you want to do is to run the 2 versions of your app side-by-side, and use YARP to route traffic to one or the other based on criteria such as URL or user properties etc?

Typically, you will run YARP as a separate process from the two versions of the app - and so the rate limiter that is built into the app will not affect traffic coming into YARP, unless you specifically include it in the YARP process.

**Reply by Tratcher** [reply]
Are you migrating by running YARP in the .NET 6 app as a fallback route?

Yarp has two layers, one that integrates with routing and a lower level one for direct proxying. The routing integrated component cannot be re-ordered. However the direct proxying component can be moved up by putting it in your own custom middleware ahead of others. The trade off here is having to do your own routing and config.
https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html

**Reply by GingerMoon** [reply]
Thank you guys for the reply.
I'd like to add more detais to my question description.
We want to switch all the traffic to the new migrated .net6 app.
Inside this new .net6 app, we use the reverse-proxy to relay those requests to old app according to the configuration.
For those requests to non-configured endpoints, the new .net6 app will serve those requests by itself.
That's why old and new apps have conflicts for rate limiting.

Please let me know if it is a typical use case for using YARP.
Meanwhile, I am checking "direct proxying component" recommended by @Tratcher and will update this discussion later.

**Reply by Tratcher** [reply]
That's the approach we're recommending over here: https://github.com/dotnet/systemweb-adapters

**Comment by GingerMoon** [comment, marked as answer]
> Are you migrating by running YARP in the .NET 6 app as a fallback route?
> 
> Yarp has two layers, one that integrates with routing and a lower level one for direct proxying. The routing integrated component cannot be re-ordered. However the direct proxying component can be moved up by putting it in your own custom middleware ahead of others. The trade off here is having to do your own routing and config. https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html

Thanks a lot @Tratcher , it works!
